### ISSUE TITLE
/api/rates takes a long time to respond

### ISSUE THREAD

**Issue by toolness** [18F/calc#519]
So we had a big spike in average response time on Tuesday:

![screen shot 2016-08-04 at 7 46 05 am](https://cloud.githubusercontent.com/assets/124687/17401142/f506a2ba-5a19-11e6-8c6f-ffbe63fe8bbf.png)

I'm not sure why this is, but looking at a single page load on my development machine with region 10 (but not schedule 70) data loaded, the obvious stand-out is the `/api/rates` request made via ajax, which takes 3.35 seconds on my system.

I tried using [`ab`](https://httpd.apache.org/docs/2.4/programs/ab.html) to benchmark and it claims we are responding to 0.41 requests per second and taking 24s per request when 10 concurrent clients are visiting `/api/rates` over 100 separate requests.

<details>
<summary>Full output of `ab -k -c 10 -n 100` on `/api/rates`</summary>
```
$ ab -k -c 10 -n 100 'http://app.calc.docker:8000/api/rates/?format=json&histogram=12&sort=current_price&query_type=match_all&min_experience=0&max_experience=45&contract-year=current&experience_range=0%2C45'
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking app.calc.docker (be patient).....done


Server Software:        WSGIServer/0.2
Server Hostname:        app.calc.docker
Server Port:            8000

Document Path:          /api/rates/?format=json&histogram=12&sort=current_price&query_type=match_all&min_experience=0&max_experience=45&contract-year=current&experience_range=0%2C45
Document Length:        70734 bytes

Concurrency Level:      10
Time taken for tests:   246.167 seconds
Complete requests:      100
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      7100400 bytes
HTML transferred:       7073400 bytes
Requests per second:    0.41 [#/sec] (mean)
Time per request:       24616.731 [ms] (mean)
Time per request:       2461.673 [ms] (mean, across all concurrent requests)
Transfer rate:          28.17 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    1   0.3      1       3
Processing: 18366 24387 2282.6  24287   29333
Waiting:    18354 24139 2248.0  24136   29160
Total:      18366 24388 2282.7  24287   29334

Percentage of the requests served within a certain time (ms)
  50%  24287
  66%  25408
  75%  26138
  80%  26528
  90%  27332
  95%  28782
  98%  29281
  99%  29334
 100%  29334 (longest request)
```
</details>

In other words, it seems like we can't handle a ton of load.  I'm wondering if we should add some basic [caching](https://docs.djangoproject.com/en/1.10/topics/cache/) to the API, and/or investigate the cause of the slowness.

**Comment by toolness** [comment]
Well, it's not the database that makes the request so slow--the SQL queries only take about 50ms.  After enabling [Django Debug Toolbar's profiling panel](http://django-debug-toolbar.readthedocs.io/en/1.4/panels.html#profiling) and visiting `/api/rates/` in my browser (I'm really appreciating django REST framework's browser-friendly views right now) it looks like most of the time is being spent in `statistics.py`:

![screen shot 2016-08-04 at 8 19 18 am](https://cloud.githubusercontent.com/assets/124687/17401569/685073b6-5a1c-11e6-906f-c7e4b6664ff2.png)

That said, I don't *think* that our production instance is actually using `statistics.py`--it's likely still using numpy, as it's running code prior to the merging of #337.  But even though numpy is likely faster than our custom python code, the fact that we're calculating aggregate statistics over tens of thousands of database rows on every request is a bit concerning.

**Comment by toolness** [comment]
<details>
  <summary>Adding some trivial caching to `/api/rates` makes things run much faster.</summary>
```diff
--- a/api/urls.py
+++ b/api/urls.py
@@ -1,9 +1,12 @@
 from django.conf.urls import patterns, url
 from api import views
 
+from django.views.decorators.cache import cache_page
+
+
 urlpatterns = patterns(
     '',
-    url(r'^rates/$', views.GetRates.as_view()),
+    url(r'^rates/$', cache_page(60 * 15)(views.GetRates.as_view())),
     url(r'^rates/csv/$', views.GetRatesCSV.as_view()),
     url(r'^search/$', views.GetAutocomplete.as_view()),
 )
```
</details>

However, we'd also need to make sure that newly-approved price lists invalidate the cache.  And we'd probably want to decide what kind of caching backend to use. And we'd want to make sure that we're using [`Vary` headers](https://docs.djangoproject.com/en/1.10/topics/cache/#using-vary-headers) appropriately.

This would also only really help with folks visiting the front page, I guess.  If lots of concurrent users were hitting CALC with very different kinds of queries, we'd probably be back at square one.

**Comment by jseppi** [comment]
Production is still using numpy. 

I haven't taken a close look yet but I wonder if we could offload those stastical calculations to the database.

**Comment by toolness** [comment]
Yeah, I was a bit surprised they weren't being done in the DB to begin with actually... I think that's definitely a possibility! (Though I'm curious if the CALC 1.0 team investigated it, and if so, if they left behind any clues in github issues/commits/PRs/etc...)

**Comment by jseppi** [comment]
Looks like Django QuerySet has StdDev: https://docs.djangoproject.com/en/1.8/ref/models/querysets/#stddev

**Comment by jseppi** [comment]
Good to close per changes in #529, @toolness?

**Comment by toolness** [comment]
Yeah! We're not 100% sure that the 8/2 slowdown was due to `/api/rates` being slow, but since this issue is about `/api/rates`, that's definitely been fixed by #529. Closing now!

**Comment by jseppi** [comment]
Yeah, see #549 for discussion of 8/2 outages.
